subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshottingFilter.java

```diff
@@ -28,19 +30,19 @@ public boolean isEmpty() {
         }
 
         @Override
-        public FileSystemSnapshotPredicate getAsSnapshotPredicate() {
+        public FileSystemSnapshotPredicate getAsSnapshotPredicate(Stat stat) {
             return (location, relativePath) -> true;
         }
 
         @Override
-        public DirectoryWalkerPredicate getAsDirectoryWalkerPredicate() {
+        public DirectoryWalkerPredicate getAsDirectoryWalkerPredicate(Stat stat) {
             return (path, name, isDirectory, attrs, relativePath) -> true;
         }
     };
 
     boolean isEmpty();
-    FileSystemSnapshotPredicate getAsSnapshotPredicate();
-    DirectoryWalkerPredicate getAsDirectoryWalkerPredicate();
+    FileSystemSnapshotPredicate getAsSnapshotPredicate(Stat stat);
```

Comment:
I think my basic idea was to keep `Stat` out of the API of `PatternSetSnapshottingFilter`. An arbitrary filter does not necessarily need it. For `PatternSetSnapshottingFilter` we need it, though, since we need to create a whole `FileTreeElement` there.

Why do you think having `Stat` in the API of `SnapshottingFilter` is better?